README.md
# ![ChatGPT-Bridge Logo](./src/assets/Logo/logo.png) **ChatGPT-Bridge**

# Tired of waiting for access to the GPT-4 API? Frustrated by the fees charged for using the GPT-API? We have the perfect solution for you!

Introducing **[ChatGPT-Driver](https://youtu.be/9fCtMJQxQ4c)**, a system comprised of two components: **ChatGPT-Bridge** (referred to as **"Bridge"**) and **[ChatGPT-Executor](https://github.com/improveTheWorld/ChatGPT-Executor)**  (referred to as **"Executor"**).
**Bridge** is a web browser plugin compatible with Microsoft Edge and Google Chrome. It offers API-like access to **ChatGPT**. When combined with **Executor** (a server application that receives and executes Windows commands), they unlock the full potential of ChatGPT without any waiting list or additional fees.

**Bridge** offers API-like access to GPT-3.5 and GPT-4 for third-party software via webSockets.
Bridge provides free API-like access to GPT-3.5. It also provides free API-like access to GPT-4 if you have a ChatGPT Plus account.
**✨ 🎉🌟 No fees are charged for the use of the API-like access!**

We are continuously striving to maintain and improve this project. If you love our work and find it useful, consider supporting us. Every small donation will go a long way in helping us to keep this project alive and free. **💌[Donate Here](https://www.paypal.com/donate/?hosted_button_id=SJTG7U2E6PC4W)💌**

To fully benefit from ChatGPT-Driver, install and set up both components.

## 🆕 Most recent Release Notes (Version 1.2.0)

- Add authentification token mechanism to use ChatGPT-Executor-1.2.0
- Support the newest ChatGPT page's html  ( ChatGPT May 24 Version)
- Add a donate link into pupup.html. If interressted you may do it directly from **[Here](https://www.paypal.com/donate/?hosted_button_id=SJTG7U2E6PC4W)**


**Note:** While ChatGPT-Driver has been significantly improved and tested, there may still be potential bugs and limitations. We appreciate your understanding and welcome any feedback to help us enhance the system.

## 🌟 Features

- 🔗 Seamlessly connects ChatGPT with third-party software
- 🌐 Utilizes WebSocket for real-time communication
- 🤖 Supports both GPT-3.5 and GPT-4
- 🆓 Free API-like access to GPT-3.5 and GPT-4 with a ChatGPT Plus account
- 📚 Enables ChatGPT to read large files that exceed the size of a single prompt

## 🔧 Installation and Usage

To utilize ChatGPT-Driver, you'll need to install both the Bridge plugin and the Executor server application.

### ChatGPT-Bridge

Follow the instructions in the [Getting Started](https://github.com/improveTheWorld/ChatGPT-Bridge#getting-started) section to install the Bridge plugin.

### ChatGPT-Executor

You can download and install ChatGPT-Executor-1.2.0 directly from [Here](https://bit.ly/46rz4zE). For more information you may visit the [Executor GitHub Repository](https://github.com/improveTheWorld/ChatGPT-Executor) and follow the instructions.
You also can download the correct version of ChatGPT-Executor which is compatible of your ChatGPT-Bridge version , from the ChatGPT-Bridge popup.
1. click on Entensions button ( on the top left corner of your web navigator) :

![popup_1.png](./src/assets/Usage/popup_1.png)

2. From displayed the entension list , select the chatGPT-Bridge, the popup will be displayed

![popup._2.png](./src/assets/Usage/popup_2.png)


here is how to display the ChatGPT-Bridge  popup :

## 🚀 Getting Started

### Prerequisites

- Microsoft Edge or Google Chrome browser
- Basic understanding of browser extensions

### Installation

1. Clone the repository:

   git clone https://github.com/improveTheWorld/ChatGPT-Bridge.git
   
2. Open your browser and navigate to extensions ( `edge://extensions` or `chrome://extensions` depending on your browser).
3. Enable "Developer Mode" in the top-right corner.
4. Click on "Load Unpacked" and select the `ChatGPT-Bridge/src` folder.
5. The ChatGPT-Bridge plugin should now be visible in the extensions list and ready for use!

## 🛠️ Usage

1. After installing the plugin, go to [ChatGPT](https://chat.openai.com/chat) website.
2. The bridge popup will be displayed at the top right corner of the webpage. You may move it using your mouse if needed.

* If the displayed popup is as below:

  ![Connecting.GIF](./src/assets/Usage/Connecting.GIF)

  `You forgot to launch the Executor or your third-party server, if applicable. Please start the Executor and wait. The Bridge will continuously poll for a connection until it is established.`
* When the connection is established, the displayed popup should be as below:

  ![Start.GIF](./src/assets/Usage/Start.GIF)

  Start a new chat and then click the start button to begin bridging between your ChatGPT AI and the third-party software (the Executor).

  When you click start with a new chat page, the Bridge plugin will empty the first prompt and send it. The first prompt is intended to teach CHATGPT the communication protocol to use with the Executor.

  Once ChatGPT gets all the initial instructions, it will ask you for its first task. Assign one and watch the magic work.

  `As an example of use, we asked ChatGPT to troubleshoot our WiFi card. We had an issue that the Windows network diagnostic tools were unable to detect. ChatGPT, after trying different approaches using our ChatGPT-Bridge system, was able to fix it for us.`[Video](https://youtu.be/9fCtMJQxQ4c)

  ---
* When you are done and want to stop using the bridge, just click the stop button:

  ![Stop.GIF](./src/assets/Usage/Stop.GIF)

3. The default WebSocket communication between the Bridge and the third-party software is done over port 8181. You can change the port in the config.json file if needed.
4. You can find the initial prompt text in the file firstPrompt.txt. You may customize it for your personal needs if you want.
5. Two modes of communication are available:

   * Wait for a complete GPT message before sending it to the third-party software (default mode).
   * Stream the message while it's being received from GPT (enable this by setting "streamingMode": true in the config.json file).
6. If you want to use only the Bridge plugin with your third-party software, your software should act as a server and listen on the correct port for a connection. Once the plugin is installed and launched, it will automatically request a connection.

**FROM VERSION 1.1.0 AND LATER**

If you are using a ChatGPT Plus account, a countdown of available message credits will be displayed once you start using GPT-4 (as only 25 messages per 3 hours are allowed).  

  ![Stop.GIF](./src/assets/Usage/CountDown.GIF)

  The countdown will also display a timer counting down until the message credits are refilled

<!-- Documentation
-------------

For more detailed information on how to use ChatGPT-Bridge, please refer to the [Wiki](https://github.com/improveTheWorld/ChatGPT-Bridge/wiki). -->

## 📧 Contributing

We welcome contributions! If you'd like to contribute, please follow these steps:

1. Fork the repository.
2. Create your feature branch (`git checkout -b feature/your-feature`).
3. Commit your changes (`git commit -am 'Add some feature'`).
4. Push to the branch (`git push origin feature/your-feature`).
5. Open a Pull Request.

Please read our [Contributing Guidelines](./CONTRIBUTING.md) for more details.

## ⚙️ Possible improvements

* See the [IMPROVEMENTS](./IMPROVEMENTS.md) file.

## 🔐 License

* This project is licensed under the Apache V2.0 for free software use - see the [LICENSE](./LICENSE-APACHE.txt) file for details.
* For commercial software use, see the [LICENSE\_NOTICE](./LICENSE_NOTICE.md) file.

## 📬 Contact

If you have any questions or suggestions, please feel free to reach out to us:

* [Tec-Net](mailto:tecnet.paris@gmail.com)

<!-- *   [Project Link](https://github.com/improveTheWorld/ChatGPT-Bridge) -->

## 🎉 Acknowledgments

* [OpenAI](https://www.openai.com/) for the [ChatGPT](https://chat.openai.com/chat) website
* [Google](https://www.google.com/chrome/) and [Microsoft](https://www.microsoft.com/en-us/edge) for providing the browser platform
* All contributors who have helped improve ChatGPT-Bridge
* The amazing open-source community for their invaluable resources and inspiration
* Our users and testers for providing essential feedback that helps us improve ChatGPT-Bridge

**Join us on this exciting journey to unlock the unlimited power and potential of ChatGPT! Stay tuned for future updates and the upcoming release of the other part of the project, which will bring even more groundbreaking features and possibilities. Together, we can revolutionize the way we interact with AI-powered language models!**


ReleaseNotes.md
# V1.2.0, What's new?
 
* Add authentification token mechanism to use ChatGPT-Executor-1.2.0,
* Support the newest ChatGPT page's html  ( ChatGPT May 24 Version),
* Add a donate link into popup.html.

src/content.js
// # File: content.js
// # Software: ChatGPT-Bridge
// # Purpose: Provide third-party software with the capability to utilize ChatGPT.
// #
// # Copyright © 2023 Tec-Net
// #
// # Dual License Notice
// #
// # For Free Software Projects:
// # This software is licensed under the Apache License, Version 2.0;
// # you may not use this file except in compliance with the License.
// # You may obtain a copy of the License at
// #
// #     http://www.apache.org/licenses/LICENSE-2.0
// #
// # For Non-Free Software Projects:
// # This software requires a commercial license. Please contact the author at
// # tecnet.paris@gmail.com to discuss licensing terms and pricing.
// #
// # Unless required by applicable law or agreed to in writing, software
// # distributed under the License is distributed on an "AS IS" BASIS,
// # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// # See the License for the specific language governing permissions and
// # limitations under the License.

let lastSentMessage = '';
let previousMessage = '';
let previousMessageTimestamp = 0;
let mostRecentMessagePollingIntervalId = null;
let reconnectionPollIntervalId = null;
let connectionStatusNotificationIntervalId = null;
let config = null;
let connectionStatus = 'connecting';
let ws = null
let firstPrompt = null;
let key = null ;
let isAuthenticated = false;

let textarea = document.querySelector('div.relative.flex > textarea');

let sendButton = null; 

function getMostRecentMessage() {
    const messageElement = document.querySelector('div.group:nth-last-child(2) div.markdown.prose');

    if (!messageElement) return;

    return messageElement.innerText.trim();

}

function checkForNewCommands() {
    if (!config) return;


    const currentMessage = getMostRecentMessage();

    // Continue reading the most recent received message until it doesn't change during the defined time
    if (currentMessage && (currentMessage !== previousMessage)) {

        //Streaming mode management: to test
        if (config.streamingMode) {
            const appendedMessage = currentMessage.slice(previousMessage.length).trim();
            //console.log(appendedMessage);
            sendMessageToWebSocketServer(appendedMessage);
        }

        //update current 
        previousMessage = currentMessage;
        previousMessageTimestamp = Date.now();

        // if (currentMessage == previousMessage) && waiting period reached
    } else if (Date.now() - previousMessageTimestamp >= config.messageCompletionTime && !config.streamingMode) {

        if (currentMessage !== lastSentMessage) {
            lastSentMessage = currentMessage;
            //console.log('Most recent received message:', completeMessage);
            sendMessageToWebSocketServer(lastSentMessage);
        }
    }
}

function adjustTextAreaHeight() {
    // Adjust the textarea height based on the content length
    const rows = Math.ceil((textarea.value.length) / 88);
    const height = rows * 24;
    textarea.style.height = height + 'px';
}


function AnswerChatGPT(message) {
    // Fetching the textarea and the sendButton with more specific selector
    textarea = document.querySelector('#prompt-textarea');
    sendButton = document.querySelector('#prompt-textarea').closest('div').querySelector('button');
    if (!textarea || !sendButton) {
        console.error('Unable to find textarea or send button.');
        return;
    }

    // Focus on the textarea
    textarea.focus();
    
    // Simulating user typing into the field with triggering input event
    setNativeValue(textarea, message);
    textarea.dispatchEvent(new Event('input', { bubbles: true }));

    adjustTextAreaHeight();

    // Enable the send button if it's disabled
    if (sendButton.hasAttribute('disabled')) {
        sendButton.removeAttribute('disabled');
    }
    
    sendButton.click();
    return ;
}

function setNativeValue(element, value) {
    let lastValue = element.value;
    element.value = value;
    let event = new Event("input", { target: element, bubbles: true });
    // React 15
    event.simulated = true;
    // React 16-17
    let tracker = element._valueTracker;
    if (tracker) {
        tracker.setValue(lastValue);
    }
    element.dispatchEvent(event);
}


function isSendError()
{
    return (document.querySelector('form.stretch div.relative > div:first-child').textContent.trim() === 'There was an error generating a responseRegenerate response');
}

function notifyConnectionStatus() {
    emitEvent(connectionStatus);
}


function encryptStringAES(message, key) {
    const iv = CryptoJS.lib.WordArray.random(128/8); // generate a random initialization vector
    const encrypted = CryptoJS.AES.encrypt(message, CryptoJS.enc.Base64.parse(key), { 
        iv: iv, 
        mode: CryptoJS.mode.CBC 
    });

    // Concatenate the random IV and the cipher text, then convert to a base64 string
    const encryptedMessage = CryptoJS.enc.Base64.stringify(iv.concat(encrypted.ciphertext));
    return encryptedMessage;
}

function extractKey(str) {
    const regex = /^KEY:(.*)$/;
    const match = str.match(regex);
    
    if (match && match.length === 2) {
      return match[1];
    }
    
    return null;
  }

  
function connectWebSocket() {
    if (connectionStatus != 'disconnected') {
        // Check if the WebSocket is already connected before creating a new connection
        if (ws &&(ws.readyState === WebSocket.OPEN || ws.readyState === WebSocket.CONNECTING) ){
            console.log("connectWebSocket : already connected or connecting !");
            return;
        }

        ws = new WebSocket(`ws://127.0.0.1:${config.communicationPort}`);

        ws.addEventListener('message', (event) => {

            const output = event.data;
            console.log("New message received :", output);
            if(config.authenticationNeeded && !isAuthenticated)
            {
                tempKey = extractKey(output);

                if(tempKey != null)
                {
                    console.log("New Key receiced : "+ tempKey);
                    key = tempKey;
                    save('funKey', key);

                }
                else if(key!= null)
                {
                    console.log("Authentication question received. Sending Answer");
                    ws.send( encryptStringAES(CryptoJS.SHA256(output).toString(),key));
                   
                }
                else
                {
                    console.error("not authentified, authentication needed, but key value is "+key);
                }

                if(key!= null)
                {
                    isAuthenticated = true;

                    if (connectionStatus != 'disconnected') {
                        connectionStatus = 'connected';
                    }   
                }
                
            }
            else
            {
                AnswerChatGPT(output);
            }
        });

        ws.addEventListener('open', (event) => {
            if( !config.authenticationNeeded && (connectionStatus != 'disconnected') )
            {
                connectionStatus = 'connected';
            }
        });

        ws.addEventListener('error', (event) => {
            if (connectionStatus != 'disconnected') {
                connectionStatus = 'connecting';
                setTimeout(() => {
                    connectWebSocket();
                }, config.reconnectInterval);
            }
        });

        ws.addEventListener('close', (event) => {
            isAuthenticated = false;
            if (connectionStatus != 'disconnected') {
                connectionStatus = 'connecting';
                setTimeout(() => {
                    connectWebSocket();
                }, config.reconnectInterval);
            }
        });
    }
}

async function loadConfig() {
    try 
    {
        const storedConfig = await getStored('config');
        if (storedConfig) 
        {
            config = storedConfig;
        } else 
        {
            const response = await fetch(chrome.runtime.getURL('config.json'));
            config = await response.json();
            await save('config',config); // Save the default config to local storage
        }
    }
    catch (error) 
    {
        console.error('Error loading config:', error);
    }
}


async function load(itemName, variable, defaultValue) 
{
    try 
    {
        const temp = await getStored(itemName);
        console.log("getStored("+itemName+") returned :", temp);
        if (temp) 
        {
            return temp;
        } 
        else
        { 
            if (defaultValue) 
            {
                await save(itemName,defaultValue); // Save the default config to local storage
            }

            return defaultValue;
        }
    } 
    catch (error) 
    {
        console.error('Error loading config:', error);
        if (defaultValue) 
        {
            await save(itemName,defaultValue); // Save the default config to local storage
        }
        
        return defaultValue;
    }
}


function getStored(itemName) {
    return new Promise((resolve, reject) => {
        chrome.storage.local.get(itemName, (result) => {
            if (chrome.runtime.lastError) {
                console.error('Error retrieving '+ itemName + ' :', chrome.runtime.lastError);
                reject(chrome.runtime.lastError);
            } else {
                console.log( itemName + " loaded: "+ result[itemName]);
                resolve(result[itemName]);
            }
        });
    });
}

function save(itemName, item ) {
    return new Promise((resolve, reject) => {
        chrome.storage.local.set({ [itemName] : item }, () => {
            if (chrome.runtime.lastError) {
                console.error('Error saving ' + itemName +' :', chrome.runtime.lastError);
                reject(chrome.runtime.lastError);
            } else {
                console.log(itemName + ' saved successfully as' + item);
                resolve();
            }
        });
    });
}

async function loadTimestamps() {
    try {
        const temp = await getStored('timestamps')

        if (temp) {
 
            timestamps = temp;

        }
    } catch (error) {
        console.error('Error loading timestamps:', error);
    }

    if (chrome.runtime.lastError) {
        reject(chrome.runtime.lastError);
    }
}

function saveTimestamps(timestamps) {
    return save('timestamps', timestamps);
}



async function loadfirstPrompt() {
    try {
        const response = await fetch(chrome.runtime.getURL('firstPrompt.txt'));
        firstPrompt = await response.text();
    } catch (error) {
        //console.error('Error loading firstPrompt:', error);
        if (chrome.runtime.lastError) {
            console.error('Error retrieving config:', chrome.runtime.lastError);
            reject(chrome.runtime.lastError);
        } 

    }

    
}

function emitEvent(name, detail = null) {
    const event = new CustomEvent(name, { detail });
    window.dispatchEvent(event);
}


function injectPopup() {

    const popupHtml = `
    <script src="https://cdnjs.cloudflare.com/ajax/libs/crypto-js/4.0.0/crypto-js.min.js"></script>

    <div id="popup-container" style="position: fixed; top: 0px; right: 10px; z-index: 9999; background-color: #f5f5f5; padding: 10px; border-radius: 5px; box-shadow: 0 1px 3px rgba(0, 0, 0, 0.2);  width: 140px;">
        <div style="display: flex; align-items: stretch; justify-content: space-between;">
            <div style="display: flex; flex-direction: column; justify-content: center; border: 1px solid #ccc; border-radius: 5px; padding: 5px; background-color: #e0e0e0;">
                <h2 id="topIcon" style="margin: 0;"><i class="fas fa-chain-broken"></i> Bridge</h2>
                <button id="startButton" style="margin-top: 5px;"><i class="fas fa-play"></i></button>
            </div>
            
            <div id="countdown-container" style="display: none; flex-direction: column; justify-content: center; border: 1px solid #ccc; border-radius: 5px; padding: 5px; background-color: #e0e0e0;">
                <div>
                    <h2 id="countDown" style="margin-right: 5px; margin-left: 5px;  display: inline;">25</h2>
                    <small id="countDownMessage" style="font-size: 9px;"></small>
                </div>
                <div>
                    <small id="timerCountDown" style="font-size: 13px;"> 00h:0min</small>
                </div>
            </div>
        </div>
    </div>`;
    
    const parser = new DOMParser();
    const popupDOM = parser.parseFromString(popupHtml, 'text/html');
    document.body.appendChild(popupDOM.querySelector('#popup-container'));

    // Inject the injected-popup.js script into the chat page
    const script = document.createElement('script');
    script.src = chrome.runtime.getURL('injected-popup.js');
    (document.head || document.documentElement).appendChild(script);

}



/////////////////////////////////////// countdown management /////////////////////////
const MAX_ALLOWED_MESSAGES_CREDIT= 25;
let allowedMessagesCount = MAX_ALLOWED_MESSAGES_CREDIT;
let timestamps = [];
 const timeLimit = 3 * 60 * 60 * 1000; // 3 hours in milliseconds


async function updateTimestamps(addNewStamps) {

    if ( await updatePendingMessagesCredit(addNewStamps))
    {
        save('timestamps',timestamps);

        timeCountDown = -1;

        if(timestamps.length > 0)
        {
            const now = new Date().getTime();
            timeCountDown = timeLimit - (now - timestamps[0])+ 1000;
        }
        emitEvent('newCountDown', { 'countDown': MAX_ALLOWED_MESSAGES_CREDIT - timestamps.length,'timerCounDown': timeCountDown  });
    }
}

async function updatePendingMessagesCredit(addNewStamps) {

    const now = new Date().getTime();
    let timestampsChanged = false;

    if (addNewStamps && (timestamps.length<MAX_ALLOWED_MESSAGES_CREDIT))
    {
        await loadTimestamps();
        timestamps.push(now);
        timestampsChanged = true;
    }

   
    // Remove debited message
    while (timestamps.length > 0 && (now - timestamps[0]) > timeLimit) {
        
        if(!timestampsChanged)
        {
            loadTimestamps();
        }
        timestamps.shift();
        timestampsChanged = true;
    }


    // Planify the next Check
    if (timestamps.length > 0) {

        const nextMessageWindow = timeLimit - (now - timestamps[0]) + 1000;

        setTimeout(() => {
            updateTimestamps(false);
        }, nextMessageWindow );
        console.log('Next check planned in :',nextMessageWindow);
    }
    return timestampsChanged;
   
}

async function onFeedbackSent() {
    if ( !isSendError() ) {
        lastSentMessage = '';
        if(isGPT4() )
        {
            //console.log('GPT 4 feedbacksent');        
        await updateTimestamps(true);
        }        
    }
}



//////////////////////////////////////// server messages management /////////////////////////////////////////////

function sendMessageToWebSocketServer(message) {
    if (ws && ws.readyState === WebSocket.OPEN) {
        ws.send(message);
    } else {
        //console.log('WebSocket is not connected or not in the open state.');
    }
}

function isNewChatPage()
{
    //  true if no  answer received yet
    let isNew = (!document.querySelector('div.group:nth-last-child(2) div.markdown.prose'))
    if(isNew)
    {
        // for each new session check the status of the account
        updateGPT_PlusAccountStatus()
    }
    return isNew;
}
function startSendingFeedback() {
       

    //If new chat session, start by send the configuration prompt
    if (isNewChatPage()) {

        if(firstPrompt)
        {
            AnswerChatGPT(firstPrompt);
        }
        else
        {
            sendMessageToWebSocketServer("_START_NEW");
        }
        
    }
    startMonitoringReceivedMessages();
}

function stopSendingFeedback() {
   
    stopMonitoringReceivedMessages();

}

/////////////////////////  Server connection Management   ///////////////////////////


function disconnectServer() {
    clearInterval(connectionStatusNotificationIntervalId);
    connectionStatus = 'disconnected';
    if (ws && (ws.readyState !== WebSocket.CLOSED) && (ws.readyState !== WebSocket.CLOSING)) {
        ws.close();
    }
}

function connectToServer() {
    connectionStatus = 'connecting';
    connectWebSocket();
    connectionStatusNotificationIntervalId = setInterval(notifyConnectionStatus, config.reconnectInterval);
}

let isChatGPT_PlusAccount = false
function updateGPT_PlusAccountStatus()
{
    if(document.querySelector('span.bg-yellow-200.text-yellow-900.py-0\\.5.px-1\\.5.text-xs.md\\:text-sm.rounded-md.uppercase'))
    {
        isChatGPT_PlusAccount = true;
    }
}

  


function isGPT4()
{
    //when the chat page is just loaded
    let titleElement = document.querySelector('.truncate > span.flex.h-6.items-center.gap-1.truncate');
    if(!titleElement){
        titleElement = document.querySelector('.flex.w-full.items-center.justify-center.gap-1.border-b');
    }
    if(!titleElement)
    {
        titleElement = document.querySelector('div[class*="dark\\:text-gray-300"]');
    }
    if(!titleElement)
    {
        return false;
    }    
    
    const regex = new RegExp(`\\b${'GPT-4'}\\b`, 'i');
    return regex.test(titleElement.innerText);
   
}


// Helper function to check if an element or any of its ancestors matches the condition
function isSendButton(element) {
    while (element) {
        if (
            element.tagName === 'BUTTON' &&
            element.parentElement.querySelector('textarea.w-full') !== null
        ) {
            return true;
        }
        element = element.parentElement;
    }
    return false;
}

// Helper function to check if an element or any of its ancestors is the desired textarea
function isTextArea(element) {
    while (element) {
        if (
            element.tagName === 'TEXTAREA' &&
            element.classList.contains('w-full')
        ) {
            return true;
        }
        element = element.parentElement;
    }
    return false;
}


/////////////////////////  Main Program  ///////////////////////////

window.addEventListener('startSendingFeedback', () => {
    console.log(' content.js: startSendingFeedback received ');
    startSendingFeedback();
});


window.addEventListener('stopSendingFeedback', () => {
    //console.log(' content.js: stopSendingFeedback received ');
    stopSendingFeedback();
});

window.addEventListener('disconnect', () => {
    //console.log(' content.js: disconnect received ');
    disconnectServer();
});

window.addEventListener('connect', () => {
    //console.log(' content.js: connect received ');
    connectToServer();
});



//Attaching the event listener to the document
document.addEventListener('keydown', (event) => {
    // Check if the event target (or any of its ancestors) is the desired textarea
    if (isTextArea(event.target)) {
        // Check if the Enter key was pressed
        if (event.key === 'Enter' ) {

            // Wait for a brief delay (e.g., 100 milliseconds)
            setTimeout(() => {
                // Check if the textarea still contains the text
                if (event.target.value !== '') {
                    // The Enter keypress was not considered as validation, handle the false Enter case here
                    //console.log('False Enter detected');
                } else {
                    
                    // The Enter keypress was considered as validation, continue with your onFeedbackSent() function
                    onFeedbackSent();
                }
            }, config.messageCompletionTime-500);
        }
    }
});

// Attaching the event listener to the document
document.addEventListener('click', (event) => {
    // Check if the clicked element (or any of its ancestors) matches the condition
    if (isSendButton(event.target)) {
        setTimeout(() => {
            onFeedbackSent();
        },config.messageCompletionTime-500);
    }
});

// Add an event listener for the 'message' event
window.addEventListener('message', (event) => {
    // Check if the message is of type 'popupReady'
    if (event.data && event.data.type === 'popupReady') {
        onPopupReady();
    }
});

function stopMonitoringReceivedMessages()
{
    sendMessageToWebSocketServer("_STOP_");
    clearInterval( mostRecentMessagePollingIntervalId);

}

function startMonitoringReceivedMessages()
{
    lastSentMessage = '';
    previousMessage = '';
    previousMessageTimestamp = 0;

    if (mostRecentMessagePollingIntervalId) {
        clearInterval(mostRecentMessagePollingIntervalId);
      }

    mostRecentMessagePollingIntervalId = setInterval(checkForNewCommands, config.pollingFrequency);
}


async function init() {
    await loadConfig();
    await loadfirstPrompt();
    key = await load('funKey', key, null)
    console.log("loaded key : " + key);
    injectPopup();
 

    //Start Monitoring
    connectToServer();
    await loadTimestamps();
    await updateTimestamps(false);
    //startMonitoringReceivedMessages();
    // Wait for the popup to be ready before emitting the event
        document.addEventListener('popupReady', function() {
            if( isGPT4() || timestamps.length >0) {
                const now = new Date().getTime();
                emitEvent('newCountDown', { 'countDown': MAX_ALLOWED_MESSAGES_CREDIT - timestamps.length,'timerCounDown':timeLimit - ( now - timestamps[0])+ 1000 });
            }
        });

    
}

init();

